paddle/fluid/framework/details/all_reduce_deps_pass.cc

```diff
@@ -48,19 +48,71 @@ std::unique_ptr<ir::Graph> AllReduceDepsPass::ApplyImpl(
   auto graph_ops = ir::FilterByNodeWrapper<OpHandleBase>(*graph);
 
   // get vars order
-  int order = 0;
   std::unordered_map<std::string, int> vars;
-  // TODO(gongwb): use graph topology sort to find the order of operators.
-  //               Note that must assert topology sort is stable
   auto& ops = Get<const std::vector<OpDesc*>>(kAllOpDescs);
-  for (auto* op_desc : ops) {
-    auto outputs = op_desc->Outputs();
-    for (auto& o_it : outputs) {
-      for (auto& v : o_it.second) {  // values
-        vars[v] = order;
+
+  // TODO(liangdun): add a config in build strategy
```

Comment:
If sortting is always better, there is no need to add a config in build strategy.  Maybe can test on more models to see the performance.

Reply:
I wonder that the original code is also one kind of topological-sort order? So which kind of topological-sort order would be the best? It seems to be an interesting question.

Reply:
I see changes in optimizer.py and here. Which change results in the performance improvements?

Reply:
If the sort in optimizer.py is removed, the order of reduction will be the order of grad op creation.
If bfs topological sorting is used, the order of reduction will be bfs topological sorting order.
Both two orders improve performance. The bfs order may be better in some case. Because the bfs order is close to the execution order when ParallelExecutor is used.

Currently, the reduction order is dictionary order.

Reply:
Can we first remove the sort in optimizer.py? not sure why that sort is needed.

We need to have more experiments for the bfs sort codes. Can this be a default for all topological sort?